inference-engine/src/mkldnn_plugin/emitters/jit_snippets_emitters.hpp

```diff
@@ -0,0 +1,637 @@
+// Copyright (C) 2020-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <ngraph/rt_info.hpp>
+#include <ngraph/variant.hpp>
+
+#include "jit_emitter.hpp"
+using namespace Xbyak;
+namespace MKLDNNPlugin {
+
+#define SNIPPETS_MAX_SNIPPETS_DIMS 7
+#define SNIPPETS_MAX_HARNESS_DIMS 5
+#define SNIPPETS_MAX_TILE_RANK 2
+#define GET_OFF(field) offsetof(jit_snippets_call_args, field)
+struct jit_snippets_call_args {
+    const void *src_ptrs[SNIPPETS_MAX_SNIPPETS_DIMS] = {};
+    void *dst_ptrs[SNIPPETS_MAX_SNIPPETS_DIMS] = {};
+};
+
+struct jit_snippets_compile_args {
+    int64_t scheduler_dims[SNIPPETS_MAX_TILE_RANK] = {};
+    int64_t scheduler_offsets[SNIPPETS_MAX_SNIPPETS_DIMS] = {};
+    int64_t data_offsets[SNIPPETS_MAX_SNIPPETS_DIMS * SNIPPETS_MAX_HARNESS_DIMS] = {};
+    std::vector<int64_t> output_dims = {};
+};
+///
+/// \brief    Kernel is the only entry point to Codogen Jit compilation. Kernel calculates appropriate data offsets,
+/// and invokes enclosed outer Tiles. Only 2d Tiles are currently supported, so the emitters should
+/// be organized in the following way:
+/// KernelEmitter {          /* entry point */
+///     TileEmitter {        /* outer tile */
+///         TileEmitter {    /* inner vector tile */
+///             ...          /* All the necessary Load/Strore/elementwise emitters */
+///         }
+///         TileEmitter {    /* inner scalar tile for tail processing */
+///             ...          /* All the necessary Load/Strore/elementwise emitters */
+///         }
+///     }
+/// }
+/// Note that Kernel params are passed directly to the emit_code(). The vector of inputs should contain 2 arguments, the
+/// output vector should be empty. Input parameters
+///
+/// \param      num_inputs       The number of the node inputs
+/// \param      num_outputs      The number of the node outputs
+///
+// Todo: Scheduler dims and offsets are currently calculated in MKLDNN Subgraph node and passed to the KernelEmitter.
+//  However, it seems more natural to calculate all the offsets right in the Kernel op, because the calculation is
+//  not device-specific. It is based only on input/output dims (which we already know) and harness num dims
+//  (which we should pass from the plugin). It seems also better to wrap the enclosed emitters in tiles in the Kernel op
+//  and avoid creating empty tiles.
+class KernelEmitter : public jit_emitter {
+public:
+    KernelEmitter(mkldnn::impl::cpu::x64::jit_generator* h, mkldnn::impl::cpu::x64::cpu_isa_t isa,
+    const std::shared_ptr<ov::Node>& n)
+    : jit_emitter(h, isa, n) {
+        const auto kernel = ov::as_type_ptr<ngraph::snippets::op::Kernel>(n);
```

Comment:
Should we check if this `n` is really `Kernel`?

Reply:
You are right, we should in case someone wants to create Kernel from other op) Thanks for noticing, exception added